pkg/cluster/base.go

```diff
@@ -1331,3 +1373,69 @@ func (p *ProviderBase) ValidateRequireSSHPrivateKey() error {
 	}
 	return nil
 }
+
+func (p *ProviderBase) ValidateAddons() error {
+	for _, plugin := range p.Enable {
+		if plugin != "explorer" {
+			_, err := common.DefaultDB.GetAddon(plugin)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (p *ProviderBase) parseDefaultTemplates() template.FuncMap {
+	templateValues := p.parseAddonTemplates()
+	return template.FuncMap{
+		"providerMetadata": func(index, metaName string) string {
+			node := templateValues[index]
+			nodeType := reflect.TypeOf(node)
+			fieldValue := ""
+			found := true
+			for i := 0; i < nodeType.NumField(); i++ {
+				field := nodeType.Field(i)
+				if v, ok := field.Tag.Lookup("json"); ok {
+					fieldName := strings.Split(v, ",")[0]
+					if strings.EqualFold(fieldName, metaName) {
+						found = true
+						if field.Type.Kind() == reflect.Slice {
+							value, ok := reflect.ValueOf(node).Field(i).Interface().([]string)
+							if ok {
+								fieldValue = value[0]
+							}
+						} else {
+							fieldValue = reflect.ValueOf(node).Field(i).String()
+						}
+						break
+					}
+
+				}
+			}
+			if !found {
+				p.Logger.Warnf("[%s] there's no metaName %s defined", p.Provider, metaName)
+			}
+			return fieldValue
+		},
+	}
+}
+
+func (p *ProviderBase) parseAddonTemplates() map[string]types.Node {
```

Comment:
I don't think this function is really needed because k3s requires all addon manifests should be the same in all master node.
And all addon manifests are only deployed into master node.

Reply:
I have changed the default template to get the first master node parameter. Since the sync.Map can't garantee the order of instances. Changed to use master node slice which is read from DB and will stay the same sequence all the time.